File: cpp/11_二进制中的1.cpp
#include <iostream>
#include <vector>
#include <stdio.h>
#include <sys/time.h>

class Solution {
public:
    // 循环移位法,注意需要转换类型
    // 时间复杂度为O(32)
    int NumberOf1(int n) {
        unsigned int tmp = n;
        int result = 0;
        while (tmp > 0) {
            if (tmp & 1 == 1) {
                result++;
            }
            tmp >>= 1;
        }
        return result;
    }

    // 依次找最后一位1,不需要转换类型
    // 时间复杂度:O(1的个数)
    int NumberOf1_easy(int n) {
        // 这种方式不需要转换符号
        int cnt = 0;
        while (n != 0) {
            cnt++;
            n &= (n - 1);
        }
        return cnt;
    }

    // 测试函数
    void test(int n) {
        int result= 0;
        struct timeval start, end;
        printf("=====\n");
        for (auto func : func_vec_) {
            gettimeofday(&start, nullptr);
            result = (this->*func)(n);
            gettimeofday(&end, nullptr);
            printf("result: %2d, times(us): %ld\n", result, end.tv_usec - start.tv_usec);
        }
    }

private:
    typedef int (Solution::*func_ptr)(int);
    std::vector<func_ptr> func_vec_ = {&Solution::NumberOf1, &Solution::NumberOf1_easy};
};

int main(int argc, char* [])
{
    Solution s;
    s.test(8);
    s.test(-7);
    s.test(-1);

    return 0;
}

File: python/11_二进制中的1.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

class Solution():
    def NumberOf1(self, n):
        """
        注意与操作0xffffffff
        这样并不是获得了一个补码
        而是获得了一个正数，这个正数与n的补码具有相同个数的1
        
        python中，对负数右移会陷入死循环，因为负数的最高位最是保证为1
        因此python中对负数进行二进制操作时，先要使用0xffffffff化为等价的正数
        """
        result = 0
        if n < 0:
            n = n & 0xffffffff
        while n > 0:
            if n & 1 == 1:
                result += 1
            n = n >> 1
        return result

    def NumberOf1_easy(self, n):
        """
        高效解法
        """
        cnt = 0
        if n < 0:
            n = n & 0xffffffff
        while n != 0:
            cnt += 1
            n = n & (n - 1)
        return cnt
            

def main():
    s = Solution()
    print(s.NumberOf1(8))
    print(s.NumberOf1(-7))

    print(s.NumberOf1_easy(8))
    print(s.NumberOf1_easy(-7))


if __name__ == "__main__":
    main()